### PR TITLE
Marking success/failed automatically clears failed downstream tasks 


### LINKED ISSUE #12485: Optionally clear downstream failed tasks when marking success
**Description**

For a DAG that looks like this:

```
A >> B
```

If A fails, B goes into "upstream_failed" state. If a user then marks A "success", B will remain in "upstream_failed" state. It will not automatically start running. This scenario often happens if the failure of A is dealt with outside of Airflow and the user does not want Airflow to run A again. But he usually expect Airflow to run B after A is marked "success".

**Use case / motivation**

After A is marked "success", its downstream tasks B should be cleared automatically and get ready to be scheduled. To avoid changing this behaviour completely, there can be a toggle button next to "Success" that lets the user decide if he wants Airflow to automatically clear downstream tasks that are in "upstream_failed" state.

### PR DESCRIPTION
**Pull request by yuqian90**
closes: #12485

---
When marking a failed task as success/failed in Graph View, this PR makes Airflow automatically clears downstream tasks that were blocked by the task after marking it.

For example, given a DagRun currently in this state, with ``run_after_loop`` and ``run_this_last`` blocked by the failed task ``runme_1``:
image

When the user marks ``runme_1`` as success, downstream tasks in ``upstream_failed`` state are automatically cleared:

The DagRun looks like this immediately after marking `runme_1` success. 
image

### CODE REVIEW

**Comment by turbaszek** [comment]
Could you please add a note in documentation (tbh not sure where)?

On side note, would it make sens to have a tooltip showing small description for all options in ti view? CC @ryanahamilton

**Comment by yuqian90** [comment]
> Could you please add a note in documentation (tbh not sure where)?
> 
> On side note, would it make sens to have a tooltip showing small description for all options in ti view? CC @ryanahamilton

Hi @turbaszek . Thanks. I've added a section in [ui.rst](https://github.com/apache/airflow/blob/dab11f2d5e0229e01b3bb75582bd3b5ba8b7f1a9/docs/apache-airflow/ui.rst#task-instance-context-menu) under "Task Instance Context Menu". I also added a note in UPDATING.md about this feature.

**Comment by RosterIn** [comment]
@yuqian90 I think this button can also apply to Mark Failed.
If the downstream task has trigger rule of one_failed / all_failed etc

**Comment by yuqian90** [comment]
> @yuqian90 I think this button can also apply to Mark Failed.
> If the downstream task has trigger rule of one_failed / all_failed etc

Hi @RosterIn  I agree it's a possible use case. However, supporting that actually requires some more work. The current PR is using this convenience method `DAG.clear()` like this. After marking the task success, the `only_failed=True` argument makes sure that task does not get cleared by this line of code here. However, to use the same code for Mark Failed means that it's going to clear the task itself (that was just marked failed). 

```python
                    subdag.clear(
                        start_date=start_date,
                        end_date=end_date,
                        include_subdags=True,
                        include_parentdag=True,
                        only_failed=True,
                        session=session,
                    )
```

To Mark Failed work work with Resume is not difficult, probably just need to add another argument to `DAG.clear()` to exclude the task itself or something like that. That requires more work and testing. 

That said, `one_failed` or `all_failed` are much less common than `all_success`. So I'm tempted to doing that in another PR.

**Comment by ashb** [comment]
Is there a case when you _wouldn't_ want this behaviour? If we can not add a new checkbox to that already monsterously complex modal I (and Ryan!) would be happier.

File: airflow/www/templates/airflow/dag.html [hunk 1]
```diff
@@ -310,6 +310,10 @@ <h4>Task Actions</h4>
                   <input type="checkbox" value="true" name="success_downstream" autocomplete="off">
                   Downstream
                 </label>
+                <label class="btn btn-default btn-sm active" title="Automatically clear blocked downstream tasks">
```
Comment (ashb) [inline airflow/www/templates/airflow/dag.html]:
```suggestion
                <label class="btn btn-default btn-sm active" title="Automatically clear downstream tasks that were in 'upstream_failed' state">
```

(If that's actually what it does?

File: airflow/www/views.py [hunk 2]
```diff
@@ -1741,16 +1742,44 @@ def _mark_task_instance_state(  # pylint: disable=too-many-arguments
...
+                )
+
+                if state == State.SUCCESS and resume:
+                    # If Resume is checked when marking success, clear downstream tasks that
+                    # are in upstream_failed state to resume them.
+
+                    # Flush the session so that the tasks marked success are reflected in the db.
+                    session.flush()
+                    subdag = dag.sub_dag(
+                        task_ids_or_regex=fr"^{task_id}$",
```
Comment (ashb) [inline airflow/www/views.py]:
```suggestion
                        task_ids_or_regex=[task_id],
```

File: airflow/www/views.py [hunk 3]
```diff
@@ -1741,16 +1742,44 @@ def _mark_task_instance_state(  # pylint: disable=too-many-arguments
...
+                    session=session,
+                )
+
+                if state == State.SUCCESS and resume:
+                    # If Resume is checked when marking success, clear downstream tasks that
+                    # are in upstream_failed state to resume them.
+
+                    # Flush the session so that the tasks marked success are reflected in the db.
+                    session.flush()
+                    subdag = dag.sub_dag(
```
Comment (ashb) [inline airflow/www/views.py]: This is the deprecated name for it

```suggestion
                    partial_dag = dag.partial_subset(
```

File: airflow/www/views.py [hunk 4]
```diff
@@ -1741,16 +1742,44 @@ def _mark_task_instance_state(  # pylint: disable=too-many-arguments
...
+
+                    end_date = execution_date if not future else None
+                    start_date = execution_date if not past else None
+
+                    subdag.clear(
+                        start_date=start_date,
+                        end_date=end_date,
+                        include_subdags=True,
+                        include_parentdag=True,
+                        only_failed=True,
```
Comment (ashb) [inline airflow/www/views.py]: This is technically going to clear FAILED as well as UPSTREAM_FAILED.

**Comment by ryanahamilton** [comment]
I agree with @ashb that adding another button wouldn't be ideal. If we _do_ need to add one, the current placement isn't quite appropriate within that button group since the other options are for task selection. Something like this would make more sense IMO:

image

**Comment by yuqian90** [comment]
Thanks @ashb  and @ryanahamilton 
I see two slightly different suggestions:
1. Do away with an extra button and make "Success + Resume" the default and only behaviour. 
2. Add the extra "Success+Resume" button like @ryanahamilton suggested, instead of adding a checkbox.

I'm fine with either. I'm tempted to go with 1 because I'm trying hard to think of a scenario where we do not want this behaviour. Does anyone have a usecase in mind where we want to mark a task success but not clear downstream failed?

**Comment by ashb** [comment]
I certainly can't think of one, 1 sounds good to me.

Question: does the (new) API use the same code path? Perhaps this could be a case where the API could have the parameter, but the UI is kept simpler.

File: airflow/www/templates/airflow/dag.html (hunk 1 above)
Comment (yuqian90) [reply airflow/www/templates/airflow/dag.html]: This change is no longer needed.

File: airflow/www/views.py (hunk 3 above)
Comment (yuqian90) [reply airflow/www/views.py]: Thanks. Done

File: airflow/www/views.py (hunk 4 above)
Comment (yuqian90) [reply airflow/www/views.py]: Thanks. Updated doc.

File: airflow/www/views.py (hunk 2 above)
Comment (yuqian90) [reply airflow/www/views.py]: Done.

**Comment by yuqian90** [comment]
> @yuqian90 I think this button can also apply to Mark Failed.
> If the downstream task has trigger rule of one_failed / all_failed etc

@RosterIn  This is now done. Marking Success/Failed will both clear downstream tasks in failed/upstream_failed state.



> I certainly can't think of one, 1 sounds good to me.
> 
> Question: does the (new) API use the same code path? Perhaps this could be a case where the API could have the parameter, but the UI is kept simpler.

HI @ash I went with option 1. The PR now looks much simpler. Please take another look. I have not looked into the API. It's not using the code path in views.py as far as I can tell. I think API changes can be done separately.

File: UPDATING.md [hunk 5]
```diff
@@ -52,6 +52,10 @@ assists users migrating to a new version.
 
 ## Master
 
+### Marking success/failed automatically clears failed downstream tasks
+
+When marking a task success/failed in Graph View, its downstream tasks that are in failed/upstream_failed state are automatically cleared.
+
```
Comment (turbaszek) [inline UPDATING.md]: @kaxil @ashb is this a 2.1 or 3.0 change? The behaviour is changing but it should not broke DAGs or pipelines (I think).

**Comment by potiuk** [comment]
Can you please rebase @yuqian90 ? We have some serious CI slow-downs (we are working on it) but this leads to some master broken more often than we would like to have. Apologies. Rebasing should fix the problem.

**Comment by yuqian90** [comment]
> Can you please rebase @yuqian90 ? We have some serious CI slow-downs (we are working on it) but this leads to some master broken more often than we would like to have. Apologies. Rebasing should fix the problem.

Thanks. Rebased and picked up this commit from upstream/master:
```
commit 8ecdef3e50d3b83901d70a13794ae6afabc4964e (upstream/master)
Author: Kaxil Naik <kaxilnaik@gmail.com>
Date:   Tue Jan 12 10:16:01 2021 +0000

    Audit Log records View should not contain link if dag_id is None (#13619)
    
    closes https://github.com/apache/airflow/issues/13602
```

**Comment by potiuk** [comment]
> Thanks. Rebased and picked up this commit from upstream/master:
> 
> ```
> commit 8ecdef3e50d3b83901d70a13794ae6afabc4964e (upstream/master)
> Author: Kaxil Naik <kaxilnaik@gmail.com>
> Date:   Tue Jan 12 10:16:01 2021 +0000
> 
>     Audit Log records View should not contain link if dag_id is None (#13619)

That too. It's the parent of this one that fixes the problem (:crossed_fingers: )

File: UPDATING.md (hunk 5 above)
Comment (kaxil) [reply UPDATING.md]: I think 2.1

File: UPDATING.md [hunk 6]
```diff
@@ -167,6 +167,10 @@ constantly.
 From Airflow 2.0.0, the scheduling decisions have been moved from
 DagFileProcessor to Scheduler, so we can keep the default a bit higher: `30`.
 
+### Marking success/failed automatically clears failed downstream tasks
```
Comment (kaxil) [inline UPDATING.md]: This section now needs moving to the top

File: airflow/models/dag.py [hunk 7]
```diff
@@ -1190,6 +1191,8 @@ def clear(
         :param visited_external_tis: A set used internally to keep track of the visited TaskInstance when
             clearing tasks across multiple DAGs linked by ExternalTaskMarker to avoid redundant work.
         :type visited_external_tis: set
+        :param exclude_task_ids: A set of task_id that should not be cleared
```
Comment (kaxil) [inline airflow/models/dag.py]:
```suggestion
        :param exclude_task_ids: A set of ``task_id``'s that should not be cleared
```

File: airflow/www/views.py [hunk 8]
```diff
@@ -1811,16 +1811,43 @@ def _mark_task_instance_state(  # pylint: disable=too-many-arguments
...
-                tasks=[task],
-                execution_date=execution_date,
-                upstream=upstream,
-                downstream=downstream,
-                future=future,
-                past=past,
-                state=state,
-                commit=True,
-            )
+            with create_session() as session:
```
Comment (kaxil) [inline airflow/www/views.py:1813]: Can you add tests for it @yuqian90

**Review by kaxil — APPROVED** [review]

**Review by ashb — COMMENTED** [review]
@yuqian90 @kaxil This need tackling in a different way -- there shouldn't be _any_ logic in the view, as that means that the Web UI and the API behave differently.

File: airflow/models/dag.py [hunk 9]
```diff
@@ -1153,6 +1153,7 @@ def clear(
         max_recursion_depth=None,
         dag_bag=None,
         visited_external_tis=None,
+        exclude_task_ids: FrozenSet[str] = frozenset({}),
```
Comment (ashb) [inline airflow/models/dag.py:1156]: This also shouldn't be a frozen set (or even _any_ set) -- but a `typing.Container[str]`

**Comment by yuqian90** [comment]
> @yuqian90 @kaxil This need tackling in a different way -- there shouldn't be _any_ logic in the view, as that means that the Web UI and the API behave differently.

Hi @ashb . I can't seem to find an API function that has similar features as `_mark_task_instance_state()` in `views.py`. If we don't put this logic in `views.py`, should i move it into `airflow.api.common.experimental.mark_tasks.set_state()` ?

**Comment by ashb** [comment]
@yuqian90  https://github.com/apache/airflow/blob/75c91b4acf1ed45d6ccf60a6e1326700233a4f05/airflow/api_connexion/endpoints/task_instance_endpoint.py#L272 is the comparable endpoint in the API isn't it?

**Comment by yuqian90** [comment]
> @yuqian90
> 
> https://github.com/apache/airflow/blob/75c91b4acf1ed45d6ccf60a6e1326700233a4f05/airflow/api_connexion/endpoints/task_instance_endpoint.py#L272
> 
> is the comparable endpoint in the API isn't it?

Thanks. That's a good point. Should we change `airflow.api.common.experimental.mark_tasks.set_state` so that we can make both `_mark_task_instance_state()` and `post_set_task_instances_state` clear downstream tasks in `upstream_failed` state when a failed task is marked success?

**Comment by ashb** [comment]
I don't mind if we don't change the behaviour of the deprecated API (and probably we shouldn't change it) but so long as the OpeAPI and UI behave the same I'm good.

**Comment by yuqian90** [comment]
> I don't mind if we don't change the behaviour of the deprecated API (and probably we shouldn't change it) but so long as the OpeAPI and UI behave the same I'm good.

Hi @ash, please see this PR that addresses your comment: https://github.com/apache/airflow/pull/16539

**Pull request merged**